### PR TITLE
Feature prune join output

### PR DESCRIPTION
**Pull request by sopel39**
FYI: @martint @dain 

Partial micro benchmark results:

```
    Partial results:
    Benchmark                                             (buildHashEnabled)  (hashColumns)  (matchRate)  (outputColumns)  Mode  Cnt    Score   Error  Units
    BenchmarkHashBuildAndJoinOperators.benchmarkJoinHash                true         bigint            2          varchar  avgt   60  157,080 ± 3,820  ms/op
    BenchmarkHashBuildAndJoinOperators.benchmarkJoinHash                true         bigint            2           bigint  avgt   60   90,912 ± 1,954  ms/op
    BenchmarkHashBuildAndJoinOperators.benchmarkJoinHash                true         bigint            2              all  avgt   60  203,645 ± 3,665  ms/op
```

`all` is a baseline. When you require just some columns the join time reduces significantly.

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/BenchmarkHashBuildAndJoinOperators.java [hunk 1]
```diff
@@ -174,20 +164,52 @@ protected void initializeBuildPages()
         @Param({"0.1", "1", "2"})
         protected double matchRate;
 
+        @Param({"varchar", "bigint", "all"})
+        protected String outputColumns;
```
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/operator/BenchmarkHashBuildAndJoinOperators.java]: I would drop "varchar" param. It doesn't tell us anything about performance changes more compared to "bigint" vs "all". Especially in the context of ubenchto, where we don't want to blob number of test cases.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 2]
```diff
@@ -197,7 +197,11 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Set<Symbol>> context)
             PlanNode left = context.rewrite(node.getLeft(), leftInputs);
             PlanNode right = context.rewrite(node.getRight(), rightInputs);
 
-            return new JoinNode(node.getId(), node.getType(), left, right, node.getCriteria(), node.getFilter(), node.getLeftHashSymbol(), node.getRightHashSymbol());
+            List<Symbol> outputSymbols = node.getOutputSymbols().stream()
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: Shouldn't there be some kind of support in `ProjectionPushDown` as well?

**Comment by pnowojski** [comment]
This will probably conflict with https://github.com/prestodb/presto/pull/6395

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: `ProjectionPushDown` pushes it through Union or Exchange only. Anyway, JoinNode is not as flexible as projections (cannot do computations, left symbols must be before right symbols)

**Comment by sopel39** [comment]
reopening

**Pull request closed without merging**